Add Comment structure with links and tags

joffrey-bion opened the pull request:
This pull request implements a `CommentParser` responsible for creating a structure `Comment` object with pure `CommentText`s,  `InlineLink`s,  and more general `InlineTag`s.

The addition of `equals` and `hashcode` methods was only useful for writing tests more easily. They could be removed by changing how the tests are made, but I believe they don't hurt, and on the contrary they may be valuable later for clients of this lib too.

A next step would be to actually access imports, so that we could return `InlineLink`s with actual `Class` objects instead of class names.

Resolves:
https://github.com/dnault/therapi-runtime-javadoc/issues/17

ST-DDT commented on therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java:
```diff
@@ -0,0 +1,110 @@
+
+    @Test
+    public void parse_mix_textAndLink() {
+        List<CommentElement> elements = CommentParser.parse("text before {@link ClassName} text after").getElements();
```
Please also add a test for a text like this:

`"Adds the given {@link Action}s to the queue."`

joffrey-bion replied:
Isn't this what this test is about? Or do you mean with no whitespace following the closing brace?

ST-DDT replied:
Yes. The main use cases usually dont break during a refactor, but these edge cases do.

joffrey-bion replied:
Done.

ST-DDT commented on therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java:
```diff
@@ -0,0 +1,84 @@
+    private static final Pattern linkRefSplitter = Pattern.compile("#");
+
+    static Comment parse(String commentText) {
+        String[] parts = inlineTagSplitter.split(commentText);
```
IMO this could be better solved with a inlineTag matcher find startIndex endIndex search. If you need an example I can write it later today.

joffrey-bion replied:
Indeed this is a better approach, it was around 3-4am, probably the reason why I used lookaheads instead !

joffrey-bion replied:
Done.

joffrey-bion commented:
I made a couple changes to improve this PR.

Lookaheads were overkill and added unnecessary complexity that required clean up. Things like "foo}bar" were broken down to `["foo}", "bar"]`, so we had to merge them as a single text part. Using a `Matcher` solved this problem the right way.

Empty/blank tags (like `{@tag}` or `{@tag   }` seem to be taken as tags by the Javadoc tool, so we should do so too. I added support for that.

Things like `foo {@ } bar` are forbidden by the Javadoc tool, so we should not worry too much about them. The `CommentParser` currently leaves them as plain text.

joffrey-bion commented:
A couple more additions:
- blank comments should not return a `Comment` object, because technically there is no comment. This would force the users to deal with the absence of comment themselves, which is not nice.
- comments should be trimmed. I can't think of a situation where people actually would want additional blank lines or space in their description. On the other hand, if the global parser for the javadoc block leaves whitespace when splitting blocks, we'd better make sure it is removed by the `CommentParser`.

For a more detailed list of use cases, please simply read the tests ;-)

dnault commented:
Thanks, nice work!
